src/main/java/com/hbm/inventory/container/ContainerBase.java

```diff
@@ -0,0 +1,67 @@
+package com.hbm.inventory.container;
+
+import net.minecraft.entity.player.EntityPlayer;
+import net.minecraft.entity.player.InventoryPlayer;
+import net.minecraft.inventory.Container;
+import net.minecraft.inventory.IInventory;
+import net.minecraft.inventory.Slot;
+import net.minecraft.item.ItemStack;
+
+/**For now, only used for stuff with filters and crates as a reference implementation,
+ * because I really needed to get the te from a container
+ * But you should very much use this to kill the giant amount of boilerplate in container classes
+ * @author 70k **/
+public class ContainerBase extends Container {
+
+    public IInventory te;
+
+    public ContainerBase (InventoryPlayer invPlayer, IInventory tedf){
+        te = tedf;
+    }
+    @Override
+    public boolean canInteractWith(EntityPlayer player) {
+        return te.isUseableByPlayer(player);
+    }
+
+    @Override
+    public ItemStack transferStackInSlot(EntityPlayer p_82846_1_, int par2) {
+        ItemStack var3 = null;
+        Slot var4 = (Slot) this.inventorySlots.get(par2);
+
+        if(var4 != null && var4.getHasStack()) {
+            ItemStack var5 = var4.getStack();
+            var3 = var5.copy();
+
+            if(par2 <= te.getSizeInventory() - 1) {
+                if(!this.mergeItemStack(var5, te.getSizeInventory(), this.inventorySlots.size(), true)) {
+                    return null;
+                }
+            } else if(!this.mergeItemStack(var5, 0, te.getSizeInventory(), false)) {
+                return null;
+            }
+
+            if(var5.stackSize == 0) {
+                var4.putStack(null);
+            } else {
+                var4.onSlotChanged();
+            }
+
+            var4.onPickupFromSlot(p_82846_1_, var5);
+        }
+
+        return var3;
+    }
+
+    public void playerInv(InventoryPlayer invPlayer, int playerInvX, int playerInvY, int playerHotbarY){
```

Comment:
I've got this implementation I did for the transporter rocket that simplifies creating quick grids. You'll want to increase the visibility from `private` to `protected` and remove the snarky comment probably lmao.

```java
	IInventory transporterInventory;

	public ContainerTransporterRocket(InventoryPlayer invPlayer, IInventory invTransporter) {
		transporterInventory = invTransporter;

		addSlots(invTransporter, 0, 8, 34, 2, 4); // Input slots
		addSlots(invTransporter, 8, 98, 34, 2, 4); // Output slots

		addSlots(invTransporter, 16, 8, 106, 1, 4); // Input Fluid ID slots
		addSlots(invTransporter, 20, 188, 90, 1, 2); // Fuel Fluid ID slots

		addSlots(invPlayer, 9, 8, 154, 3, 9); // Player inventory
		addSlots(invPlayer, 0, 8, 212, 1, 9); // Player hotbar
	}

	// I'm gonna make a farken helper function for this shit, why was it done the old way for 9 whole ass years?
	private void addSlots(IInventory inv, int from, int x, int y, int rows, int cols) {
		int slotSize = 18;
		for(int row = 0; row < rows; row++) {
			for(int col = 0; col < cols; col++) {
				this.addSlotToContainer(new Slot(inv, col + row * cols + from, x + col * slotSize, y + row * slotSize));
			}
		}
	}
```

Reply:
seems excellent, I'll add tomorrow